File: frontend/src/lib/calendarUtils.ts
/** Shared calendar day helpers (local keys + UTC ISO from API). */

export function pad2(n: number) {
  return String(n).padStart(2, "0");
}

export function localDayKey(y: number, monthIndex: number, day: number) {
  return `${y}-${pad2(monthIndex + 1)}-${pad2(day)}`;
}

export function calendarDayKeyFromIso(iso: string) {
  const d = new Date(iso);
  return `${d.getUTCFullYear()}-${pad2(d.getUTCMonth() + 1)}-${pad2(d.getUTCDate())}`;
}

export function daysBetweenKeys(fromKey: string, toKey: string) {
  const parse = (k: string) => {
    const [y, m, d] = k.split("-").map(Number);
    return new Date(y, m - 1, d);
  };
  return Math.floor((parse(toKey).getTime() - parse(fromKey).getTime()) / 86400000);
}

/** Whole days past the original deadline (0 if not overdue). */
export function taskDelayDays(deadlineIso: string) {
  const dueKey = calendarDayKeyFromIso(deadlineIso);
  const now = new Date();
  const todayKey = localDayKey(now.getFullYear(), now.getMonth(), now.getDate());
  const days = daysBetweenKeys(dueKey, todayKey);
  return days > 0 ? days : 0;
}

/** Calendar day an open deliverable appears on — overdue tasks roll forward to today. */
export function runwayCalendarDayKey(deadlineIso: string, status: string) {
  if (status === "COMPLETED") return calendarDayKeyFromIso(deadlineIso);
  if (taskDelayDays(deadlineIso) > 0) {
    const now = new Date();
    return localDayKey(now.getFullYear(), now.getMonth(), now.getDate());
  }
  return calendarDayKeyFromIso(deadlineIso);
}

export function runwayStatusIsDelayed(task: { deadline: string; status: string }) {
  return task.status !== "COMPLETED" && taskDelayDays(task.deadline) > 0;
}

export function runwayStatusLabel(task: { deadline: string; status: string }) {
  const delay = taskDelayDays(task.deadline);
  if (delay > 0 && task.status !== "COMPLETED") return `DELAYED · ${delay}d`;
  switch (task.status) {
    case "IN_PROGRESS":
      return "IN PROGRESS";
    case "PENDING":
      return "PENDING";
    case "DELAYED":
      return "DELAYED";
    case "COMPLETED":
      return "COMPLETED";
    default:
      return task.status.replaceAll("_", " ");
  }
}

export function groupOpenTasksByDeadlineDay<T extends { deadline: string; status: string }>(tasks: T[]) {
  const map = new Map<string, T[]>();
  for (const t of tasks) {
    if (t.status === "COMPLETED") continue;
    const key = runwayCalendarDayKey(t.deadline, t.status);
    map.set(key, [...(map.get(key) ?? []), t]);
  }
  return map;
}

export function monthRangeIso(y: number, monthIndex: number) {
  const last = new Date(y, monthIndex + 1, 0);
  return { from: localDayKey(y, monthIndex, 1), to: localDayKey(y, monthIndex, last.getDate()) };
}

export const WEEKDAYS = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"] as const;

export function formatDisplayDate(isoOrKey: string) {
  const key = isoOrKey.includes("T") ? calendarDayKeyFromIso(isoOrKey) : isoOrKey;
  const [y, m, d] = key.split("-").map(Number);
  return new Date(y, m - 1, d).toLocaleDateString();
}

export function taskTypeLabel(taskType: string) {
  return taskType.replaceAll("_", " ").toLowerCase().replace(/(^|\s)\w/g, (m) => m.toUpperCase());
}

/** Standard deliverable offsets from event (shoot) day. */
export const DELIVERABLE_DEADLINE_DAYS: { label: string; days: number }[] = [
  { label: "Hard drives", days: 60 },
  { label: "Sneak peak", days: 7 },
  { label: "Full set photos", days: 20 },
  { label: "Cinematic video", days: 20 },
  { label: "Reels", days: 20 },
  { label: "Traditional video", days: 45 },
  { label: "Album design", days: 45 },
  { label: "Album print", days: 60 },
];

export function deadlineSummaryText() {
  return DELIVERABLE_DEADLINE_DAYS.map((d) => `${d.label} +${d.days}d`).join(" · ");
}

File: mobile/lib/core/calendar_utils.dart
String pad2(int n) => n.toString().padLeft(2, '0');

String localDayKey(DateTime date) =>
    '${date.year}-${pad2(date.month)}-${pad2(date.day)}';

/// Stable `YYYY-MM-DD` for calendar grouping (matches backend `parseDayUtc`).
String shootDayKey(String value) {
  final s = value.trim();
  if (RegExp(r'^\d{4}-\d{2}-\d{2}$').hasMatch(s)) return s;
  final d = DateTime.parse(s).toUtc();
  return '${d.year}-${pad2(d.month)}-${pad2(d.day)}';
}

/// UTC calendar day from an ISO timestamp (task deadlines, etc.).
String calendarDayKeyFromIso(String iso) => shootDayKey(iso);

String formatFriendlyDay(String isoOrKey, {bool includeYear = false}) {
  final key = isoOrKey.contains('T') ? calendarDayKeyFromIso(isoOrKey) : isoOrKey;
  final parts = key.split('-').map(int.parse).toList();
  final date = DateTime(parts[0], parts[1], parts[2]);
  const months = [
    'January', 'February', 'March', 'April', 'May', 'June',
    'July', 'August', 'September', 'October', 'November', 'December',
  ];
  final base = '${date.day} ${months[date.month - 1]}';
  return includeYear ? '$base ${date.year}' : base;
}

int daysBetweenKeys(String fromKey, String toKey) {
  DateTime parse(String k) {
    final p = k.split('-').map(int.parse).toList();
    return DateTime(p[0], p[1], p[2]);
  }
  return parse(toKey).difference(parse(fromKey)).inDays;
}

const weekdayLabels = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];

/// Sunday-start week containing [date].
DateTime startOfWeek(DateTime date) {
  final local = DateTime(date.year, date.month, date.day);
  return local.subtract(Duration(days: local.weekday % 7));
}

List<DateTime> weekDaysFrom(DateTime weekStart) {
  return List.generate(7, (i) => weekStart.add(Duration(days: i)));
}

const deliverableDeadlineRules = [
  ('Hard drives', 60),
  ('Sneak peak', 7),
  ('Full set photos', 20),
  ('Cinematic video', 20),
  ('Reels', 20),
  ('Traditional video', 45),
  ('Album design', 45),
  ('Album print', 60),
];

Map<String, List<T>> groupOpenTasksByDeadlineDay<T>(
  Iterable<T> tasks,
  String Function(T) deadlineOf,
  String Function(T) statusOf,
) {
  final map = <String, List<T>>{};
  for (final t in tasks) {
    if (statusOf(t) == 'COMPLETED') continue;
    final key = runwayCalendarDayKey(
      deadlineIso: deadlineOf(t),
      status: statusOf(t),
    );
    map.putIfAbsent(key, () => []).add(t);
  }
  return map;
}

/// Calendar day an open deliverable appears on — overdue tasks roll forward to today.
String runwayCalendarDayKey({required String deadlineIso, required String status}) {
  if (status == 'COMPLETED') return calendarDayKeyFromIso(deadlineIso);
  if (taskDelayDays(deadlineIso) > 0) return localDayKey(DateTime.now());
  return calendarDayKeyFromIso(deadlineIso);
}

/// Whole days past the original deadline (0 if not overdue).
int taskDelayDays(String deadlineIso) {
  final dueKey = calendarDayKeyFromIso(deadlineIso);
  final todayKey = localDayKey(DateTime.now());
  final days = daysBetweenKeys(dueKey, todayKey);
  return days > 0 ? days : 0;
}

bool isTaskOverdue(String deadlineIso, {String? status}) {
  if (status == 'COMPLETED') return false;
  return taskDelayDays(deadlineIso) > 0;
}
